lib/Service/PhishingDetectionService.php

```diff
@@ -0,0 +1,190 @@
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+namespace OCA\Mail\Service;
+
+use DateTime;
+use Horde_Mime_Headers;
+use OCA\Mail\AddressList;
+use OCA\Mail\Contracts\ITrustedSenderService;
+
+class PhishingDetectionService {
+
+
+	private ContactsIntegration $contactIntegration;
+
+	private ITrustedSenderService $trustedSenderService;
+
+
+	private bool $warn = false;
+
+
+	public function __construct(ContactsIntegration $contactIntegration, ITrustedSenderService $trustedSenderService) {
+		$this->contactIntegration = $contactIntegration;
+		$this->trustedSenderService = $trustedSenderService;
+	}
+
+	private function checkDatePass(string $date): array {
+		$now = new DateTime();
+		$dt = new DateTime($date);
+		if($dt > $now) {
+			$this->warn = true;
+			return ["check" => false , "message" => "Sent date is in the future"];
+		}
+		return ["check" => $dt < $now ];
+	}
+
+	private function replyToCheckPass(string $fromEmail, ?string $replyToEmail): array {
+		if(!(isset($replyToEmail))) {
+			return ["check" => true];
+		}
+		if($replyToEmail !== $fromEmail) {
+			$this->warn = true;
+		}
+		return ["check" => false , "message" => "Reply-To email: ${$replyToEmail} is different from the sender email: ${$fromEmail}"];
+	}
+
+	private function customEmailCheck(string $fromEmail, ?string $customEmail): array {
+		if(!(isset($customEmail))) {
+			return ["check" => true];
+		}
+		if($customEmail !== $fromEmail) {
+			$this->warn = true;
+		}
+		return ["check" => false , "message" => "Sender is using a custom email: ${$customEmail} instead of the sender email: ${$fromEmail}"] ;
+	}
+
+
+	private function contactsCheckPass(string $fn, string $email):array {
+		$emailInContacts = false;
+		$emails = "";
+		$contacts = $this->contactIntegration->getContactsWithName($fn, true);
+		foreach ($contacts as $contact) {
+			foreach ($contact['email'] as $contactEmail) {
+				$emailInContacts = true;
+				$emails .= $contactEmail.",";
+				if ($contactEmail === $email) {
+					return ["check" => true];
+				}
+			}
+		}
+		if ($emailInContacts) {
+			$this->warn = true;
+			return ["check" => false, "message" => "Sender email: ${$email} is not in the contacts list, but the sender name: ${$fn} is in the contacts list with the following emails: ${$emails}"];
+		}
+		return ["check" => true];
+	}
+
+	private function checkTrusted(string $uid, string $email): array {
+		$domain = explode('@', $email)[1];
+		$trusted = $this->trustedSenderService->isTrusted($uid, $email) || $this->trustedSenderService->isTrusted($uid, $domain);
+
+		if(!$trusted) {
+			$this->warn = true;
+			return ["check" => false, "message" => "Sender email: ${$email} is not trusted"];
+		}
+		return ["check" => true];
+	}
+
+	private function isLink(string $text): bool {
+		$pattern = '/^(https?:\/\/|www\.|[a-zA-Z0-9-]+\.[a-zA-Z]{2,})/i';
```

Comment:
+1 for not using a custom Regexp.

But if I understand correctly this method isn't supposed to match only fully qualified URLs, but all text that looks roughly like a URL (as part of a link text).

If I'm correct then `FILTER_VALIDATE_URL` would be too strict because it doesn't accept e.g. `cloud.nextcloud.com/p/something`. Additionally it doesn't accept URLs containing non-ASCII characters.

I'd suggest to use a library to find URLs in text. [The Horde lib "Text_Filter" contains a class](https://dev.horde.org/api/master/lib/Text_Filter/classes/Horde_Text_Filter_Linkurls.html) that links all found URLs, maybe some methods of that can be used? 

Also maybe the method could be renamed to something more telling (e.g. "textLooksLikeALink()"), and get a comment about its purpose?

Reply:
You could use the original source, yes. But a library might bring the benefit that possible changes (there's a list of TLDs in there, which may change) are tended for by multiple people/projects – I'm unsure about the Horde libs in this regard, though.